File: src/main/java/com/cxxcxx/zinecraft/core/biome/NationBiomePlacements.java
package com.cxxcxx.zinecraft.core.biome;

import com.cxxcxx.zinecraft.api.world.dimension.DimensionBiome;
import net.minecraft.resources.ResourceKey;
import net.minecraft.world.level.biome.Biome;
import net.minecraft.world.level.biome.Climate;
import net.minecraft.world.level.biome.Climate.ParameterPoint;

import java.util.List;

public final class NationBiomePlacements {
  public static final NationBiomePlacements INSTANCE = new NationBiomePlacements();
  public static final List<DimensionBiome> ALL;

  static {
    DimensionBiome[] dimensionBiomes = new DimensionBiome[]{
        INSTANCE.placement(ModBiome.AEGIR_ABYSSAL_SEA, -0.35F, 0.65F, -1.05F, 0.35F, 0.0F, -0.75F),
        INSTANCE.placement(ModBiome.BOLIVAR_PLAIN, 0.55F, 0.0F, 0.1F, 0.55F, 0.0F, 0.0F),
        INSTANCE.placement(ModBiome.HIGASHI_SHADOW_RIFT, -0.25F, 0.35F, 0.35F, -0.8F, 0.0F, -0.75F),
        INSTANCE.placement(ModBiome.DURIN_UNDERGROUND_GARDEN, 0.35F, 0.65F, 0.55F, -0.4F, 1.0F, -0.75F),
        INSTANCE.placement(ModBiome.COLUMBIA_SANDSTONE_WILDS, 0.9F, -0.75F, 0.9F, -0.25F, 0.0F, 0.8F),
        INSTANCE.placement(ModBiome.KAZIMIERZ_KNIGHTLAND, 0.0F, -0.35F, 0.35F, 0.55F, 0.0F, -0.2F),
        INSTANCE.placement(ModBiome.KAZDEL_SCARRED_WASTES, 0.0F, -0.85F, 0.9F, -0.8F, 0.0F, -0.55F),
        INSTANCE.placement(ModBiome.LATERANO_HOLY_FIELDS, 0.45F, 0.0F, 0.1F, 0.35F, 0.0F, 0.3F),
        INSTANCE.placement(ModBiome.LEITHANIEN_TWILIGHT_FOREST, -0.3F, 0.45F, 0.35F, -0.4F, 0.0F, 0.55F),
        INSTANCE.placement(ModBiome.RIM_BILLITON_MINING_BADLANDS, 0.85F, -0.45F, 0.9F, -0.75F, 0.0F, 0.9F),
        INSTANCE.placement(ModBiome.MINOS_SUNLIT_HILLS, 0.45F, -0.25F, 0.1F, -0.2F, 0.0F, 0.55F),
        INSTANCE.placement(ModBiome.SARGON_ROCKY_DESERT, 0.95F, -0.9F, 0.65F, 0.05F, 0.0F, -0.1F),
        INSTANCE.placement(ModBiome.SAMI_FROZEN_FOREST, -0.85F, 0.4F, 0.85F, -0.45F, 0.0F, 0.0F),
        INSTANCE.placement(ModBiome.VICTORIA_MISTY_HIGHLANDS, -0.2F, 0.7F, 0.4F, 0.0F, 0.0F, 0.65F),
        INSTANCE.placement(ModBiome.URSUS_FROZEN_STEPPE, -0.75F, -0.3F, 0.65F, 0.5F, 0.0F, -0.25F),
        INSTANCE.placement(ModBiome.KJERAG_SNOWY_PEAKS, -0.95F, 0.35F, 0.95F, -0.85F, 0.0F, 0.95F),
        INSTANCE.placement(ModBiome.SIRACUSA_RAINY_WOODLAND, 0.05F, 0.75F, 0.15F, 0.35F, 0.0F, -0.65F),
        INSTANCE.placement(ModBiome.YAN_MOUNTAIN_GROVE, 0.05F, 0.4F, 0.45F, -0.2F, 0.0F, 0.8F),
        INSTANCE.placement(ModBiome.IBERIA_SALT_DELTA, 0.5F, 0.65F, -0.2F, 0.8F, 0.0F, -0.75F),
        INSTANCE.placement(ModBiome.TERRA_CATASTROPHE_ZONE, 0.8F, -0.8F, 0.5F, -0.5F, 0.0F, 0.2F)
    };
    ALL = java.util.List.of(dimensionBiomes);
  }

  private DimensionBiome placement(
      ResourceKey<Biome> biome, float temperature, float humidity, float continentalness, float erosion, float depth, float weirdness
  ) {
    ParameterPoint parameterPoint = Climate.parameters(temperature, humidity, continentalness, erosion, depth, weirdness, 0.0F);
    return new DimensionBiome(biome, parameterPoint);
  }
}

File: src/main/kotlin/com/cxxcxx/zinecraft/core/biome/NationBiomePlacements.kt
package com.cxxcxx.zinecraft.core.biome

import com.cxxcxx.zinecraft.api.world.dimension.DimensionBiome
import net.minecraft.world.level.biome.Climate

/**
 * 泰拉维度的多噪声气候点。
 *
 * 点位覆盖海洋、海岸、内陆、山峰与地下深度；最近点分区会填满整个泰拉维度，同时不会向主世界注入任何国家群系。
 */
internal object NationBiomePlacements {
  val ALL = listOf(
    placement(NationBiomes.AEGIR_ABYSSAL_SEA, -0.35f, 0.65f, -1.05f, 0.35f, 0.0f, -0.75f),
    placement(NationBiomes.BOLIVAR_PLAIN, 0.55f, 0.0f, 0.10f, 0.55f, 0.0f, 0.0f),
    placement(NationBiomes.HIGASHI_SHADOW_RIFT, -0.25f, 0.35f, 0.35f, -0.80f, 0.0f, -0.75f),
    placement(NationBiomes.DURIN_UNDERGROUND_GARDEN, 0.35f, 0.65f, 0.55f, -0.40f, 1.0f, -0.75f),
    placement(NationBiomes.COLUMBIA_SANDSTONE_WILDS, 0.90f, -0.75f, 0.90f, -0.25f, 0.0f, 0.80f),
    placement(NationBiomes.KAZIMIERZ_KNIGHTLAND, 0.0f, -0.35f, 0.35f, 0.55f, 0.0f, -0.20f),
    placement(NationBiomes.KAZDEL_SCARRED_WASTES, 0.0f, -0.85f, 0.90f, -0.80f, 0.0f, -0.55f),
    placement(NationBiomes.LATERANO_HOLY_FIELDS, 0.45f, 0.0f, 0.10f, 0.35f, 0.0f, 0.30f),
    placement(NationBiomes.LEITHANIEN_TWILIGHT_FOREST, -0.30f, 0.45f, 0.35f, -0.40f, 0.0f, 0.55f),
    placement(NationBiomes.RIM_BILLITON_MINING_BADLANDS, 0.85f, -0.45f, 0.90f, -0.75f, 0.0f, 0.90f),
    placement(NationBiomes.MINOS_SUNLIT_HILLS, 0.45f, -0.25f, 0.10f, -0.20f, 0.0f, 0.55f),
    placement(NationBiomes.SARGON_ROCKY_DESERT, 0.95f, -0.90f, 0.65f, 0.05f, 0.0f, -0.10f),
    placement(NationBiomes.SAMI_FROZEN_FOREST, -0.85f, 0.40f, 0.85f, -0.45f, 0.0f, 0.0f),
    placement(NationBiomes.VICTORIA_MISTY_HIGHLANDS, -0.20f, 0.70f, 0.40f, 0.0f, 0.0f, 0.65f),
    placement(NationBiomes.URSUS_FROZEN_STEPPE, -0.75f, -0.30f, 0.65f, 0.50f, 0.0f, -0.25f),
    placement(NationBiomes.KJERAG_SNOWY_PEAKS, -0.95f, 0.35f, 0.95f, -0.85f, 0.0f, 0.95f),
    placement(NationBiomes.SIRACUSA_RAINY_WOODLAND, 0.05f, 0.75f, 0.15f, 0.35f, 0.0f, -0.65f),
    placement(NationBiomes.YAN_MOUNTAIN_GROVE, 0.05f, 0.40f, 0.45f, -0.20f, 0.0f, 0.80f),
    placement(NationBiomes.IBERIA_SALT_DELTA, 0.50f, 0.65f, -0.20f, 0.80f, 0.0f, -0.75f)
  )

  private fun placement(
    biome: net.minecraft.resources.ResourceKey<net.minecraft.world.level.biome.Biome>,
    temperature: Float,
    humidity: Float,
    continentalness: Float,
    erosion: Float,
    depth: Float,
    weirdness: Float
  ) = DimensionBiome(
    biome,
    Climate.parameters(temperature, humidity, continentalness, erosion, depth, weirdness, 0.0f)
  )
}
